test/e2e/azure_securitygroups.go

```diff
@@ -0,0 +1,188 @@
+			securityGroup, err := securityGroupsClient.Get(ctx, azureCluster.Spec.ResourceGroup, expectedSubnet.SecurityGroup.Name, "")
+			g.Expect(err).NotTo(HaveOccurred())
+
+			securityRules, err := securityRulesClient.List(ctx, azureCluster.Spec.ResourceGroup, *securityGroup.Name)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			var expectedSecurityRuleNames []string
+			for _, expectedSecurityRule := range expectedSubnet.SecurityGroup.SecurityRules {
+				expectedSecurityRuleNames = append(expectedSecurityRuleNames, expectedSecurityRule.Name)
+			}
+
+			for _, securityRule := range securityRules.Values() {
+				g.Expect(expectedSecurityRuleNames).To(ContainElement(*securityRule.Name))
+			}
+		}
+	}
+
+	Byf("Creating subnets for the %s cluster", input.ClusterName)
+	testSubnet := infrav1.SubnetSpec{
+		SubnetClassSpec: infrav1.SubnetClassSpec{
+			Name: "test-subnet",
+			Role: infrav1.SubnetNode,
+		},
+		SecurityGroup: infrav1.SecurityGroup{
+			Name: "test-security-group",
+			SecurityGroupClass: infrav1.SecurityGroupClass{
+				SecurityRules: []infrav1.SecurityRule{
+					testSecurityRule,
+				},
+			},
+		},
+	}
+	originalSubnets := azureCluster.Spec.NetworkSpec.Subnets
+	expectedSubnets = originalSubnets
+	expectedSubnets = append(expectedSubnets, testSubnet)
+	Eventually(func(g Gomega) {
+		g.Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(azureCluster), azureCluster)).To(Succeed())
+		azureCluster.Spec.NetworkSpec.Subnets = expectedSubnets
+		g.Expect(mgmtClient.Update(ctx, azureCluster)).To(Succeed())
+	}, inputGetter().WaitForUpdate...).Should(Succeed())
+	Eventually(checkSubnets, input.WaitForUpdate...).Should(Succeed())
+
+	By("Creating new security rule for the subnet")
+	Expect(len(expectedSubnets)).To(Not(Equal(0)))
+	testSubnet.SecurityGroup.SecurityRules = infrav1.SecurityRules{testSecurityRule, testSecurityRule2}
+	expectedSubnets = originalSubnets
+	expectedSubnets = append(expectedSubnets, testSubnet)
+	Eventually(func(g Gomega) {
+		g.Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(azureCluster), azureCluster)).To(Succeed())
+		azureCluster.Spec.NetworkSpec.Subnets = expectedSubnets
+		g.Expect(mgmtClient.Update(ctx, azureCluster)).To(Succeed())
+	}, inputGetter().WaitForUpdate...).Should(Succeed())
+	Eventually(checkSubnets, input.WaitForUpdate...).Should(Succeed())
+
+	By("Deleting security rule from the subnet")
+	Expect(len(expectedSubnets)).To(Not(Equal(0)))
+	testSubnet.SecurityGroup.SecurityRules = infrav1.SecurityRules{testSecurityRule2}
+	expectedSubnets = originalSubnets
+	expectedSubnets = append(expectedSubnets, testSubnet)
+	Eventually(func(g Gomega) {
+		g.Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(azureCluster), azureCluster)).To(Succeed())
+		azureCluster.Spec.NetworkSpec.Subnets = expectedSubnets
+		g.Expect(mgmtClient.Update(ctx, azureCluster)).To(Succeed())
+	}, inputGetter().WaitForUpdate...).Should(Succeed())
+	Eventually(checkSubnets, input.WaitForUpdate...).Should(Succeed())
```

Comment:
is the test checking that the actual rule gets deleted in Azure anywhere? I might be missing something but it seems that it's only checking that the k8s object got edited (which doesn't prove that the bug this PR is fixing is fixed AFAICT)

Reply:
ah nvm I did miss something, I missed the `checkSubnets` function :)

Reply:
This test is using a security rules client to fetch the security rules from Azure, and ensuring that each rule from Azure is specified in the spec. The logic is here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3591/files/135eb65f7db7bae4f40bab9abc4de7b9563c4505#diff-2bd0374df27acbc5890001fc0592a2399588fa52905e3b72fc428af089764bd1R118